platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc

```diff
@@ -40,11 +40,85 @@ The previous step will help you identify potential problems by issuing deprecati
+
+* Resolvable configurations should not have their resolution strategy, dependencies, hierarchy, etc., modified after they have been resolved.
+* Consumable configurations should not have their dependencies, hierarchy, attributes, etc., modified after they have been published or consumed as a variant.
+* Dependency scope configurations should not have their dependencies, constraints, etc., modified after a configuration that extends from them are observed.
+
+Gradle already throws an exception for many of these circumstances, but in some cases, it does not.
+As of Gradle 9.0, all these cases will result in an error.
+After a configuration of any type has been observed, it should be considered immutable.
+This validation covers the following properties of a configuration:
+
+* Resolution Strategy
+* Dependencies
+* Constraints
+* Exclude Rules
+* Artifacts
+* Role (consumable, resolvable, dependency scope)
+* Hierarchy (`extendsFrom`)
+* Others (Transitive, Visible)
+
+In many cases, this deprecation is caused by mutating a configuration in a link:{javadocPath}/org/gradle/api/artifacts/ResolvableDependencies.html#beforeResolve-org.gradle.api.Action-[`beforeResolve`] hook.
+This hook is only executed after a configuration is fully resolved but not when it is partially resolved for computing task dependencies.
+
+Consider the following code that showcases the deprecated behavior:
+
+=====
+[.multi-language-sample]
+======
+.build.gradle.kts
+[source,kotlin]
+----
+plugins {
+    id("java-library")
+}
+
+configurations.runtimeClasspath {
+    // `beforeResolve` is called after the configuration is resolved for task dependencies,
+    // but before it is fully resolved during artifact resolution.
+    // Configurations should not be mutated in this hook
+    incoming.beforeResolve {
+        if (allDependencies.none { it.group == "com" && it.name == "foo" }) {
+            configurations.implementation.get().dependencies.add(project.dependencies.create("com:foo:1.0"))
+        }
+    }
+}
+
+tasks.register("resolve") {
+    val conf = configurations["runtimeClasspath"]
+
+    // Wire build dependencies
+    dependsOn(conf)
+
+    // Resolve dependencies
+    doLast {
+        assert(conf.files.map { it.name } == listOf("foo-1.0.jar"))
+    }
+}
+----
+======
+=====
+
+For the following cases, consider the suggested options for replacing mutations within a `beforeResolve` hook:
+
+* **Adding dependencies**: Use link:{javadocPath}/org/gradle/api/artifacts/Configuration.html#withDependencies-org.gradle.api.Action-[`withDependencies`].
+* **Changing dependency versions**: Use dependency constraints. (See <<using_preferred_versions, Declaring preferred version constraints>>)
+* **Adding excludes**: Use <<component_metadata_rules.adoc#sec:component_metadata_rules,Component Metadata Rules>>
+* **Roles**: Configuration roles should be set upon creation and not changed afterward.
```

Comment:
Should this link to the new APIs?

Reply:
There was a release notes entry for these, but we don't really have the new APIs documented anywhere.

We could link to the release notes of _another_ version of Gradle but that seems weird (since `notes.md` is overwritten each Gradle version)